subsys/net/ip/tcp.c

```diff
@@ -2639,117 +2639,138 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
+					 * ACK so add one to new len
+					 */
+					new_len++;
 				}
 			}
 
 			/*
-			 * Only if our FIN flag has been acknowledged and all the data has been
-			 * received, it can go to the TIME_WAIT state
+			 * FIN1:
+			 * Acknowledge path and sequence path are independent, treat them that way
+			 * The for combinations and their destination states
+			 * -   & -   -> TCP_FIN_WAIT_1
+			 * FIN & -   -> TCP_CLOSING
+			 * -   & ACK -> TCP_FIN_WAIT_2
+			 * FIN & ACK -> TCP_TIME_WAIT
 			 */
-			if (FL(&fl, ==, (FIN | ACK),
-			    (th_ack(th) == conn->seq) && (th_seq(th) == conn->ack))) {
-				tcp_send_timer_cancel(conn);
-				conn_ack(conn, + 1);
-				tcp_out(conn, ACK);
-				next = TCP_TIME_WAIT;
-				NET_DBG("FIN acknowledged, going to TIME_WAIT_STATE");
-				verdict = NET_OK;
-			} else if (th &&
-				   (FL(&fl, ==, FIN, th_seq(th) == conn->ack) ||
-				    FL(&fl, ==, (FIN | ACK), th_seq(th) == conn->ack))) {
-				tcp_send_timer_cancel(conn);
-				conn_ack(conn, + 1);
-				NET_DBG("FIN not yet acknowleged, going to CLOSING state");
-				if (th_ack(th) != conn->seq) {
-					/* Not acknowledged the FIN flag yet, so resend */
-					tcp_out_ext(conn, (FIN | ACK), NULL, conn->seq - 1);
-				} else {
-					tcp_out(conn, ACK);
-				}
-				next = TCP_CLOSING;
-				verdict = NET_OK;
-			} else if (th && FL(&fl, ==, ACK, th_ack(th) == conn->seq)) {
-				tcp_send_timer_cancel(conn);
-				next = TCP_FIN_WAIT_2;
-				verdict = NET_OK;
-				NET_DBG("FIN acknowledged, going to FIN_WAIT_2 state seq %u, ack %u"
-					, conn->seq, conn->ack);
 
-				/* Acknowledge any received data */
-				if (len > 0) {
-					tcp_out(conn, ACK);
+			/*
+			 * FIN2:
+			 * Only FIN is relevant in this state, as FIN has already acknowledged
+			 * -   -> TCP_FIN_WAIT_2
+			 * FIN -> TCP_TIME_WAIT
+			 */
+
+			/*
+			 * Closing:
+			 * Our FIN has to be acknowledged
+			 * -   -> TCP_CLOSING
+			 * ACK -> TCP_TIME_WAIT
+			 */
+			if (FL(&fl, &, ACK, th_ack(th) == conn->seq)) {
+				if (conn->state == TCP_FIN_WAIT_1) {
+					NET_DBG("FIN acknowledged, going to FIN_WAIT_2 "
+						"state seq %u, ack %u"
+						, conn->seq, conn->ack);
+					tcp_send_timer_cancel(conn);
+					next = TCP_FIN_WAIT_2;
 				}
-			} else if (th) {
-				NET_DBG("Staying in FIN_WAIT_1 flags 0x%x state seq %u, ack %u",
-					fl, conn->seq, conn->ack);
-				if (len > 0) {
+				if (conn->state == TCP_CLOSING) {
+					NET_DBG("FIN acknowledged, going to TIME WAIT "
+						"state seq %u, ack %u"
+						, conn->seq, conn->ack);
 					tcp_send_timer_cancel(conn);
-					/* Send out a duplicate ACK, with the pending FIN flag */
-					tcp_out(conn, FIN | ACK);
+					next = TCP_TIME_WAIT;
 				}
 				verdict = NET_OK;
 			}
-		}
-		break;
-	case TCP_FIN_WAIT_2:
-		/* Acknowledge but drop any data, but subtract any duplicate data */
-		if (th) {
-			if (len > 0) {
-				int32_t new_len = len - net_tcp_seq_cmp(conn->ack, th_seq(th));
+			if ((next == TCP_FIN_WAIT_2) ||
```

Comment:
What about `TCP_FIN_WAIT_1` state? Shouldn't we reject new data in that state as well? RFC 9293 [3.6.1](https://www.rfc-editor.org/rfc/rfc9293#section-3.6.1):
> A host MAY implement a "half-duplex" TCP close sequence, so that an application that has called CLOSE cannot continue to read data from the connection (MAY-1).

I understand that this means we should be sending RST as soon as we close the connection on our side (no need to wait for ACK for our FIN here?)

Reply:
@rlubos
I interpreted it so that we wait for ACK for our FIN normally, but should the new data be arriving from the host, we add RST (`...while received data is still pending in the TCP connection, or if new data is received after CLOSE is called, its TCP implementation SHOULD send a RST to show that data was lost (SHLD-3)...`)


But according to RFC 9293 [3.5.3 Reset Processing](https://www.rfc-editor.org/rfc/rfc9293#name-reset-processing)](https://www.rfc-editor.org/rfc/rfc9293#section-3.5.3) receiver goes to the CLOSED state, so we do not have to ACK new data, instead we can keep sending RST.

This way at least there is no connection leak, but we better be sure that our RST has arrived, otherwise we might have a leak at the host.

But better me graceful and sync properly, just add RST to show we do not process incoming data anymore.

Reply:
I was wondering about this behavior as well, on one hand you want to reset directly, but in FIN1, there is an inflight FIN flag. If we reset right away, we will "overwrite" the inflight FIN flag. The zephyr stack only accepts in sequence RST messages for instance.
This was my rationale for not resetting in FIN1 when presented with new data, but instead acking the last accepted data.

Reply:
@rlubos @ssharks 

Here https://www.rfc-editor.org/rfc/rfc9293#section-3.5.2 it says - 
The side of a connection issuing a reset should enter the TIME-WAIT state, as this generally helps to reduce the load on busy servers for reasons described in [[70](https://www.rfc-editor.org/rfc/rfc9293#FTY99)].

So I'd say once we get data we can not process, instead of ACK-ing them, we issue RST and go to TIME-WAIT (in which we should probably keep sending RST???)

Reply:
I modified the code to also reset from FIN1. I can see the point of going to TIME_WAIT after a reset. I think this is out of scope for this PR as it affects all the places where a RST packet is generated and I think also the sequence number checking for incoming RST packet should be broadened in my opinion to work well.
Is it worth creating a new issue for this? @AndreyDodonov-EH maybe you can submit a PR for this.

Reply:
After thinking on it a bit I implemented at least the transition to TIME-WAIT from the FIN states when a resettable packet arrives. Fixing this revealed some problems with the timer, but I think I fixed it correct now.
The further I dig, the more surprised I'm that it worked reasonable in the past.